camel/functions/search_functions.py

```diff
@@ -270,40 +332,73 @@ def summarize_text(text: str, query: str) -> str:
     return response
 
 
-def search_google_and_summarize(query: str) -> str:
-    r"""Search webs for information. Given a query, this function will use
-    the Google search engine to search for related information from the
-    internet, and then return a summarized answer.
+def continue_search(query: str, answer: str) -> bool:
+    r"""Ask LLM whether to continue search or not.
 
     Args:
         query (str): Question you want to be answered.
+        answer (str): Answer to the query.
 
     Returns:
-        str: Summarized information from webs.
+        bool: True if the user want to continue search, False otherwise.
     """
-    # Google search will return a list of urls
-    responses = search_google(query)
+    prompt = TextPrompt("Do you think the ANSWER can answer the QUERY? "
+                        "Use only 'yes' or 'no' to answer\n"
+                        "===== QUERY ====={query}\n\n"
+                        "===== ANSWER ====={answer}")
+    prompt = prompt.format(query=query, answer=answer)
+    reply = prompt_single_step_agent(prompt)
+    if "yes" in str(reply).lower():
+        return False
+    return True
+
+
+def search_and_summarize(query: str, search_func) -> str:
+    """
+    Search the web for information using a specified search function
+    and summarize the results.
+
+    Args:
+        query (str): Question you want to be answered.
+        search_func (Callable): A function that takes a query
+        and returns search results.
+
+    Returns:
+        str: Summarized information from the web.
+    """
+    responses = search_func(query)
     for item in responses:
         if "url" in item:
             url = item.get("url")
-            # Extract text
-            text = text_extract_from_web(str(url))
-            # Using chatgpt summarise text
-            answer = summarize_text(text, query)
-
-            # Let chatgpt decide whether to continue search or not
-            prompt = TextPrompt(
-                '''Do you think the answer: {answer} can answer the query:
-                {query}. Use only 'yes' or 'no' to answer.''')
-            prompt = prompt.format(answer=answer, query=query)
-            reply = prompt_single_step_agent(prompt)
-            if "yes" in str(reply).lower():
+            try:
+                extracted_text = text_extract_from_web(str(url))
+            except Exception:
+                continue
```

Comment:
We use continue to handle exceptions because we don't want a single error to stop our entire search process. Imagine if we're extracting text from a web page and something goes wrong, like the network connection is down or there's an issue with the page format. We don't want to give up on the whole search because of that. So, we use continue to skip over the problematic part and keep trying other search results. This way, we increase our chances of finding a good answer. Plus, in our search_func, we can specify the maximum number of results to return, ensuring we make the most of the content we get.